File: Helpers/src/main/java/com/dg/controls/ButtonEx.java
package com.dg.controls;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Paint;
import android.graphics.Typeface;
import android.util.AttributeSet;

import com.dg.R;
import com.dg.helpers.FontHelper;

import java.util.Locale;

/**
 * Created by Daniel Cohen Gindi (danielgindi@gmail.com)
 */

public class ButtonEx extends android.widget.Button
{
    public ButtonEx(Context context)
    {
        super(context);
    }

    public ButtonEx(Context context, AttributeSet attrs)
    {
        super(context, attrs);
        setCustomFontFamily(context, attrs);
    }

    public ButtonEx(Context context, AttributeSet attrs, int defStyle)
    {
        super(context, attrs, defStyle);
        setCustomFontFamily(context, attrs);
    }

    static String _namespace;
    private void setCustomFontFamily(Context context, AttributeSet attrs)
    {
        String fontFamily = null;
        TypedArray styledAttributes = context.obtainStyledAttributes(attrs, R.styleable.CustomFontFamily);
        try
        {
            fontFamily = styledAttributes.getString(R.styleable.CustomFontFamily_customFontFamily);
        }
        finally
        {
            styledAttributes.recycle();
        }

        if (fontFamily != null && !fontFamily.isEmpty())
        {
            setPaintFlags(this.getPaintFlags() | Paint.SUBPIXEL_TEXT_FLAG | Paint.LINEAR_TEXT_FLAG);
            setCustomFont(context, fontFamily);
        }
    }

    public boolean setCustomFont(Context context, String fontFamily)
    {
        Typeface typeface = FontHelper.sharedInstance(context).getFont(fontFamily);
        if (typeface != null)
        {
            setTypeface(typeface);
            return true;
        }
        else
        {
            return false;
        }
    }
}

File: Helpers/src/main/java/com/dg/controls/ButtonEx.kt
package com.dg.controls

import android.content.Context
import android.content.res.TypedArray
import android.graphics.Paint
import android.graphics.Typeface
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatButton
import com.dg.R
import com.dg.helpers.FontHelper

@Suppress("unused")
class ButtonEx : AppCompatButton
{
    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
    {
        setCustomFontFamily(context, attrs)
    }

    constructor(context: Context,
                attrs: AttributeSet,
                defStyle: Int) : super(context, attrs, defStyle)
    {
        setCustomFontFamily(context, attrs)
    }

    private fun setCustomFontFamily(context: Context, attrs: AttributeSet)
    {
        if (isInEditMode)
        {
            return
        }

        val fontFamily: String?
        var styledAttributes: TypedArray? = null
        try
        {
            styledAttributes = context.obtainStyledAttributes(attrs, R.styleable.CustomFontFamily)
            fontFamily = styledAttributes!!.getString(R.styleable.CustomFontFamily_customFontFamily)
        }
        finally
        {
            styledAttributes?.recycle()
        }

        if (fontFamily != null && !fontFamily.isEmpty())
        {
            paintFlags = this.paintFlags or Paint.SUBPIXEL_TEXT_FLAG or Paint.LINEAR_TEXT_FLAG
            setCustomFont(fontFamily)
        }
    }

    @Deprecated("This version was a mistake", replaceWith = ReplaceWith("setCustomFont(fontFamily)"))
    fun setCustomFont(context: Context, fontFamily: String): Boolean
    {
        return setCustomFont(fontFamily)
    }

    fun setCustomFont(fontFamily: String): Boolean
    {
        val typeface: Typeface? = FontHelper.getFont(context, fontFamily)
        return if (typeface != null)
        {
            setTypeface(typeface)
            true
        }
        else
        {
            false
        }
    }
}
